### PR TITLE
Add support for pushing image with unknown digest

### PR DESCRIPTION
**Pull request by umohnani8**
Add support to push image with unknown digest and no tag to a registry. This will be used by farm build to construct a list of multi arch builds after pushing images built on each node.

### CODE REVIEW

**Comment by umohnani8** [comment]
Pulled in the changes from https://github.com/containers/podman/pull/20486 and addressed the latest round of reviews and added tests.

@mtrmac had a question on https://github.com/containers/podman/pull/20486#discussion_r1409758076, why are we deleting the first entry if it matches `ref.ref.Name()`? While testing it, we end up with this string array, which seems wrong to me https://github.com/containers/image/pull/2209/files#diff-92c905926b8c790b46f469e3f93dad938055b87977b1b3016057c4e8b0b3baa4R206

File: docker/docker_transport_test.go [hunk 1]
```diff
@@ -43,17 +44,22 @@ func TestParseReference(t *testing.T) {
...
 		// A github.com/distribution/reference value can have a tag and a digest at the same time!
 		// The docker/distribution API does not really support that (we can’t ask for an image with a specific
 		// tag and digest), so fail.  This MAY be accepted in the future.
-		{"//busybox:latest" + sha256digest, ""},                                    // Both tag and digest
-		{"//docker.io/library/busybox:latest", "docker.io/library/busybox:latest"}, // All implied values explicitly specified
-		{"//UPPERCASEISINVALID", ""},                                               // Invalid input
+		{"//busybox:latest" + sha256digest, "", false},                                     // Both tag and digest
+		{"//docker.io/library/busybox:latest", "docker.io/library/busybox:latest", false},  // All implied values explicitly specified
+		{"//UPPERCASEISINVALID", "", false},                                                // Invalid input
+		{"//busybox" + UnknownDigestSuffix, "docker.io/library/busybox", true},             // UnknownDigest suffix
```
Comment (mtrmac) [inline docker/docker_transport_test.go:61]: (Non-blocking: throughout the tests, hard-coding the `@@…@@` string, instead of referring to the same variable the primary code uses, would ensure the tests detect if we ever accidentally changed the variable value.)

File: docker/docker_transport_test.go [hunk 2]
```diff
@@ -103,6 +110,25 @@ func TestNewReference(t *testing.T) {
 	assert.Error(t, err)
 }
 
+var ValidReferenceUnknownDigestTestCases = []struct{ input, dockerRef, stringWithinTransport string }{
```
Comment (mtrmac) [inline docker/docker_transport_test.go:113]:
```suggestion
var validReferenceUnknownDigestTestCases = []struct{ input, dockerRef, stringWithinTransport string }{
```
please.

Or, alternatively (weakly preferred), add these to the existing `validReferenceTestCases`, and add a new `isUnknownDigest` field — that way we only need one simple loop in the three tests which use these tests case arrays.

File: docker/docker_transport_test.go [hunk 3]
```diff
@@ -103,6 +110,25 @@ func TestNewReference(t *testing.T) {
...
+		parsed, err := reference.ParseNormalizedNamed(in)
+		require.NoError(t, err)
+		ref, err := NewReferenceUnknownDigest(parsed)
+		require.NoError(t, err, c.input)
+		dockerRef, ok := ref.(dockerReference)
+		require.True(t, ok, c.input)
+		assert.Equal(t, c.dockerRef, dockerRef.ref.String(), c.input)
+		assert.Equal(t, true, dockerRef.isUnknownDigest)
+	}
+}
```
Comment (mtrmac) [inline docker/docker_transport_test.go:130]: Also test that tagged/digested inputs are rejected.

File: docker/docker_transport_test.go [hunk 4]
```diff
@@ -121,6 +147,18 @@ func TestReferenceStringWithinTransport(t *testing.T) {
 		stringRef2 := ref2.StringWithinTransport()
 		assert.Equal(t, stringRef, stringRef2, c.input)
 	}
+
+	for _, c := range ValidReferenceUnknownDigestTestCases {
```
Comment (mtrmac) [inline docker/docker_transport_test.go:151]: Isn’t this exactly the same code as for `validReferenceTestCases`? If so, there could just be one loop over `append(slices.Clone(validReferenceTestCases), validReferenceUnknownDigestTestCases…)`

File: docker/docker_transport_test.go [hunk 5]
```diff
@@ -131,13 +169,25 @@ func TestReferenceDockerReference(t *testing.T) {
 		require.NotNil(t, dockerRef, c.input)
 		assert.Equal(t, c.dockerRef, dockerRef.String(), c.input)
 	}
+
+	for _, c := range ValidReferenceUnknownDigestTestCases {
```
Comment (mtrmac) [inline docker/docker_transport_test.go:173]: Isn’t this exactly the same code as for `validReferenceTestCases`? If so, there could just be one loop over `append(slices.Clone(validReferenceTestCases), validReferenceUnknownDigestTestCases…)`

File: docker/docker_transport_test.go [hunk 6]
```diff
@@ -150,6 +200,14 @@ func TestReferencePolicyConfigurationNamespaces(t *testing.T) {
 		"docker.io",
 		"*.io",
 	}, ref.PolicyConfigurationNamespaces())
+
+	ref, err = ParseReference("//busybox" + UnknownDigestSuffix)
```
Comment (mtrmac) [inline docker/docker_transport_test.go:204]: If there are 3 cases in the `switch`, they should all be covered by these tests … it’s probably easier to remove the `switch`.

File: docker/docker_transport_test.go [hunk 7]
```diff
@@ -163,6 +221,16 @@ func TestReferenceNewImage(t *testing.T) {
 	})
 	require.NoError(t, err)
 	defer img.Close()
+
+	ref, err = ParseReference("//quay.io/libpod/busybox" + UnknownDigestSuffix)
+	require.NoError(t, err)
+	_, err = ref.NewImage(context.Background(), &types.SystemContext{
```
Comment (mtrmac) [inline docker/docker_transport_test.go:227]: (Absolutely non-blocking: Use a single variable to define this `SystemContext` instead of copy&pasting? Also elsewhere. This works just fine.)

File: docker/docker_transport_test.go [hunk 8]
```diff
@@ -181,6 +255,13 @@ func TestReferenceNewImageDestination(t *testing.T) {
...
+
+	ref, err = ParseReference("//quay.io/libpod/busybox" + UnknownDigestSuffix)
+	require.NoError(t, err)
+	dest2, err := ref.NewImageDestination(context.Background(),
+		&types.SystemContext{RegistriesDirPath: "/this/does/not/exist", DockerPerHostCertDirPath: "/this/does/not/exist"})
+	require.NoError(t, err)
+	defer dest2.Close()
 }
 
 func TestReferenceTagOrDigest(t *testing.T) {
```
Comment (mtrmac) [inline docker/docker_transport_test.go:267]: Also test the `isUnknownDigest` code path here

**Review by mtrmac — COMMENTED** [review]
Thanks! Looks very good overall.

- docs/containers-transports.5.md needs to be updated
- `GetDigest` should also reject `isUnknownDigest` references

File: docker/docker_image_src.go [hunk 9]
```diff
@@ -48,7 +48,13 @@ type dockerImageSource struct {
 
 // newImageSource creates a new ImageSource for the specified image reference.
 // The caller must call .Close() on the returned ImageSource.
+// The caller must ensure !ref.isUnknownDigest.
 func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	// dockerImageSource.logicalRef && dockerImageSource.physicalRef are !isUnknownDigest
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_image_src.go:53]: Fixed

File: docker/docker_image_src.go [hunk 10]
```diff
@@ -48,7 +48,13 @@ type dockerImageSource struct {
...
+	if ref.isUnknownDigest {
+		return nil, fmt.Errorf("Docker reference without a tag or digest is not supported for newImageSource")
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_image_src.go:55]: Fixed

File: docker/docker_transport.go [hunk 11]
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/containers/image/v5/types"
 )
 
+const UnknownDigestSuffix = "@@unknown-digest@@"
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:15]: Fixed. 
Added docs on this, but please let me know if you think it needs to be documented somewhere else

File: docker/docker_transport.go [hunk 12]
```diff
@@ -43,31 +45,51 @@ func (t dockerTransport) ValidatePolicyConfigurationScope(scope string) error {
...
+		refString = strings.TrimSuffix(refString, UnknownDigestSuffix)
+	}
 	ref, err := reference.ParseNormalizedNamed(strings.TrimPrefix(refString, "//"))
 	if err != nil {
 		return nil, err
 	}
+
+	if unknownDigest {
+		if !reference.IsNameOnly(ref) {
+			return nil, fmt.Errorf("DockerReference %q has unknown digest set but it contains either a tag or digest", ref.Name())
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:70]: Fixed

[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:70]: Fixed

File: docker/docker_transport.go [hunk 13]
```diff
@@ -43,31 +45,51 @@ func (t dockerTransport) ValidatePolicyConfigurationScope(scope string) error {
...
 	return NewReference(ref)
 }
 
 // NewReference returns a Docker reference for a named reference. The reference must satisfy !reference.IsNameOnly().
 func NewReference(ref reference.Named) (types.ImageReference, error) {
-	return newReference(ref)
+	return newReference(ref, false)
+}
+
+// NewReferenceUnknownDigest returns a Docker reference for a named reference. The reference must satisfy reference.IsNameOnly()
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:84]: Fixed

File: docker/docker_transport.go [hunk 14]
```diff
@@ -43,31 +45,51 @@ func (t dockerTransport) ValidatePolicyConfigurationScope(scope string) error {
...
+	return newReference(ref, true)
 }
 
 // newReference returns a dockerReference for a named reference.
-func newReference(ref reference.Named) (dockerReference, error) {
-	if reference.IsNameOnly(ref) {
-		return dockerReference{}, fmt.Errorf("Docker reference %s has neither a tag nor a digest", reference.FamiliarString(ref))
+func newReference(ref reference.Named, unknownDigest bool) (dockerReference, error) {
+	if (reference.IsNameOnly(ref) && !unknownDigest) || (!reference.IsNameOnly(ref) && unknownDigest) {
+		return dockerReference{}, fmt.Errorf("Docker reference %s is not for an unknown digest case; tag or digest is needed", reference.FamiliarString(ref))
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:92]: Fixed

File: docker/docker_transport.go [hunk 15]
```diff
@@ -126,7 +156,22 @@ func (ref dockerReference) PolicyConfigurationIdentity() string {
...
+	if ref.isUnknownDigest {
+		switch len(namespaces) {
+		case 0:
+			return namespaces
+		case 1:
+			if namespaces[0] == ref.ref.Name() {
+				return []string{}
+			}
+		default:
+			if namespaces[0] == ref.ref.Name() {
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:169]: Fixed

File: docker/docker_transport_test.go (hunk 1 above)
Comment (umohnani8) [reply docker/docker_transport_test.go:61]: Ack, fixed

File: docker/docker_transport_test.go (hunk 2 above)
Comment (umohnani8) [reply docker/docker_transport_test.go:113]: Fixed

File: docker/docker_transport_test.go (hunk 6 above)
Comment (umohnani8) [reply docker/docker_transport_test.go:204]: Fixed

File: docker/docker_transport_test.go (hunk 7 above)
Comment (umohnani8) [reply docker/docker_transport_test.go:227]: Fixed

File: docker/docker_transport_test.go (hunk 8 above)
Comment (umohnani8) [reply docker/docker_transport_test.go:267]: Fixed

File: docs/containers-transports.5.md [hunk 16]
```diff
@@ -43,6 +43,7 @@ The containers-registries.conf(5) further allows for configuring various setting
 Note that a _docker-reference_ has the following format: `name[:tag|@digest]`.
 While the docker transport does not support both a tag and a digest at the same time some formats like containers-storage do.
 Digests can also be used in an image destination as long as the manifest matches the provided digest.
+The docker transport supports pushing images without a tag or digest to a registry when the image name is suffixed with `@@unknown-digest@@`. Images without a tag or digest can be used to create manifest lists after the images have been pushed.
 The digest of images can be explored with skopeo-inspect(1).
 If `name` does not contain a slash, it is treated as `docker.io/library/name`.
```
Comment (mtrmac) [inline docs/containers-transports.5.md:48]: This is getting long and hard to parse. A paragraph break above this line?

File: docs/containers-transports.5.md [hunk 17]
```diff
@@ -43,6 +43,7 @@ The containers-registries.conf(5) further allows for configuring various setting
 Note that a _docker-reference_ has the following format: `name[:tag|@digest]`.
 While the docker transport does not support both a tag and a digest at the same time some formats like containers-storage do.
 Digests can also be used in an image destination as long as the manifest matches the provided digest.
+The docker transport supports pushing images without a tag or digest to a registry when the image name is suffixed with `@@unknown-digest@@`. Images without a tag or digest can be used to create manifest lists after the images have been pushed.
 The digest of images can be explored with skopeo-inspect(1).
```
Comment (mtrmac) [inline docs/containers-transports.5.md:47]: … or, if the destination uses the **@unknown-digest@** syntax, the writer of the manifest must provide it some other way.

File: docker/docker_image.go [hunk 18]
```diff
@@ -123,6 +123,9 @@ func GetDigest(ctx context.Context, sys *types.SystemContext, ref types.ImageRef
 	if !ok {
 		return "", errors.New("ref must be a dockerReference")
 	}
+	if dr.isUnknownDigest {
+		return "", fmt.Errorf("docker: reference %q is for unknown digest case; cannot get digest", dr.ref.Name())
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_image.go]: Fixed

File: docker/docker_transport_test.go [hunk 19]
```diff
@@ -43,17 +45,22 @@ func TestParseReference(t *testing.T) {
...
 		// The docker/distribution API does not really support that (we can’t ask for an image with a specific
 		// tag and digest), so fail.  This MAY be accepted in the future.
-		{"//busybox:latest" + sha256digest, ""},                                    // Both tag and digest
-		{"//docker.io/library/busybox:latest", "docker.io/library/busybox:latest"}, // All implied values explicitly specified
-		{"//UPPERCASEISINVALID", ""},                                               // Invalid input
+		{"//busybox:latest" + sha256digest, "", false},                                         // Both tag and digest
+		{"//docker.io/library/busybox:latest", "docker.io/library/busybox:latest", false},      // All implied values explicitly specified
+		{"//UPPERCASEISINVALID", "", false},                                                    // Invalid input
+		{"//busybox" + unknownDigestSuffixTest, "docker.io/library/busybox", true},             // UnknownDigest suffix
+		{"//example.com/ns/busybox" + unknownDigestSuffixTest, "example.com/ns/busybox", true}, // UnknownDigest with registry/repo
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport_test.go:63]: Fixed

File: docs/containers-transports.5.md (hunk 16 above)
Comment (umohnani8) [reply docs/containers-transports.5.md:50]: Fixed

File: docs/containers-transports.5.md [hunk 20]
```diff
@@ -43,6 +43,7 @@ The containers-registries.conf(5) further allows for configuring various setting
 Note that a _docker-reference_ has the following format: `name[:tag|@digest]`.
 While the docker transport does not support both a tag and a digest at the same time some formats like containers-storage do.
 Digests can also be used in an image destination as long as the manifest matches the provided digest.
+The docker transport supports pushing images without a tag or digest to a registry when the image name is suffixed with `@@unknown-digest@@`. Images without a tag or digest can be used to create manifest lists after the images have been pushed.
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply docs/containers-transports.5.md]: Fixed

File: docs/containers-transports.5.md (hunk 17 above)
Comment (umohnani8) [reply docs/containers-transports.5.md:48]: Do you want me to add this as well to the docs? I fixed up the docs a bit, let me know if you want this added as well.

File: docker/docker_transport.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (umohnani8) [reply docker/docker_transport.go:18]: Got it, fixed

File: docs/containers-transports.5.md [hunk 21]
```diff
@@ -43,7 +43,10 @@ The containers-registries.conf(5) further allows for configuring various setting
 Note that a _docker-reference_ has the following format: `name[:tag|@digest]`.
 While the docker transport does not support both a tag and a digest at the same time some formats like containers-storage do.
 Digests can also be used in an image destination as long as the manifest matches the provided digest.
+
+The docker transport supports pushing images without a tag or digest to a registry when the image name is suffixed with **@@unknown-digest@@**. The `name`**@@unknown-digest** reference format cannot be used with a reference that has a tag or digest.
```
Comment (mtrmac) [inline docs/containers-transports.5.md:47]: Expressing this as `name`**@@unknown-digest** expresses the difference between literals and variables in HTML output, but not in man pages; in man pages (with the go-md2man we use), both render as bold.

Italic (`_name_`) renders in man pages as underline, that makes the distinction work in both output formats.

Also, please update the `name[:tag|@digest]` above to express the difference between literals and variables (and [ | ] meta-syntax).

File: docs/containers-transports.5.md (hunk 17 above)
Comment (mtrmac) [reply docs/containers-transports.5.md:48]: I would weakly prefer this to be documented _somehow_, because I think it’s not immediately obvious how to use `@@unknown-digest@@` (or destinations with a specified digest).

OTOH doing that clearly, and without misleading most users who don’t need to deal with this, seems hard to do.

Meh, we can assume the current text is good enough (at least good enough to find this PR :) ) and we can improve it based on user feedback or proposed PRs.

Comment (umohnani8) [reply docs/containers-transports.5.md:48]: Sounds good, I am also not sure how to add this clearly. Will update based on any user feedback in the future.

File: docs/containers-transports.5.md [hunk 22]
```diff
@@ -40,10 +40,13 @@ By default, uses the authorization state in `$XDG_RUNTIME_DIR/containers/auth.js
 If the authorization state is not found there, `$HOME/.docker/config.json` is checked, which is set using docker-login(1).
 The containers-registries.conf(5) further allows for configuring various settings of a registry.
 
-Note that a _docker-reference_ has the following format: `name[:tag|@digest]`.
+Note that a _docker-reference_ has the following format: _name_[_:tag_ | _@digest_].
```
Comment (mtrmac) [inline docs/containers-transports.5.md:43]:
```suggestion
Note that a _docker-reference_ has the following format: _name_[**:**_tag_ | **@**_digest_].
```

File: docs/containers-transports.5.md [hunk 23]
```diff
@@ -40,10 +40,13 @@ By default, uses the authorization state in `$XDG_RUNTIME_DIR/containers/auth.js
...
 While the docker transport does not support both a tag and a digest at the same time some formats like containers-storage do.
 Digests can also be used in an image destination as long as the manifest matches the provided digest.
+
+The docker transport supports pushing images without a tag or digest to a registry when the image name is suffixed with **@@unknown-digest@@**. The _name_**@@unknown-digest** reference format cannot be used with a reference that has a tag or digest.
```
Comment (mtrmac) [inline docs/containers-transports.5.md:47]:
```suggestion
The docker transport supports pushing images without a tag or digest to a registry when the image name is suffixed with **@@unknown-digest@@**. The _name_**@@unknown-digest@@** reference format cannot be used with a reference that has a tag or digest.
```

**Review by mtrmac — COMMENTED** [review]
Thanks!

Lasts touch-ups, feel free to merge afterwards.

File: docs/containers-transports.5.md (hunk 23 above)
Comment (umohnani8) [reply docs/containers-transports.5.md]: Fixed

File: docs/containers-transports.5.md (hunk 22 above)
Comment (umohnani8) [reply docs/containers-transports.5.md]: Fixed
